server/aesdsocket.c

```diff
@@ -272,122 +510,112 @@ int main(int argc, char* argv[])
+
+            /* create socket node for each connection */
+            data_ptr = (socket_node_t *)malloc(sizeof(socket_node_t));
+            if (NULL == data_ptr)
             {
-                packet_complete = true;
+                syslog(LOG_PERROR, "malloc: %s", strerror(errno));
+                status = FAILURE;
+                goto exit;
             }
-        } while (!packet_complete);
 
-        packet_complete = false;
-
-        /* seek the file fd to start of file to read contents */
-        off_t offset = lseek(file_fd, 0, SEEK_SET);
-        if (FAILURE == offset)
-        {
-            syslog(LOG_PERROR, "lseek: %s", strerror(errno));
-            close(file_fd);
-            close(connection_fd);
-            close(socket_fd);
-            return FAILURE;
+            data_ptr->connection_fd = connection_fd;
+            data_ptr->thread_complete_success = false;
+            data_ptr->thread_mutex = &thread_mutex;
+            /* create thread for each connection */
+            if (SUCCESS != pthread_create(&data_ptr->thread_id, NULL, recv_and_send_thread, data_ptr))
+            {
+                syslog(LOG_PERROR, "pthread_create: %s", strerror(errno));
+                free(data_ptr);
+                data_ptr = NULL;
+                status = FAILURE;
+                goto exit;
+            } 
+            SLIST_INSERT_HEAD(&head, data_ptr, node_count);
         }
-        /* read file contents till EOF */
-        int read_bytes = 0;
-        int send_bytes = 0;
-        do
+        /* check whether thread exited if yes, join thread and remove from socket list */
+        data_ptr = NULL;
+        SLIST_FOREACH_SAFE(data_ptr, &head, node_count, data_ptr_temp)
         {
-            memset(buffer, 0, MAX_BUFF_LEN);
-            read_bytes = read(file_fd, buffer, MAX_BUFF_LEN);
-            if (read_bytes > 0)
+            if (data_ptr->thread_complete_success == true)
             {
-                /* send file data to client */
-                send_bytes = send(connection_fd, buffer, read_bytes, 0);
-                if (send_bytes != read_bytes)
-                {
-                    syslog(LOG_PERROR, "send: %s", strerror(errno));
-                    close(file_fd);
-                    close(connection_fd);
-                    close(socket_fd);
-                    return FAILURE;
-                }
+                syslog(LOG_INFO, "1 Joined thread id: %ld", data_ptr->thread_id);
+                pthread_join(data_ptr->thread_id, NULL);
+                SLIST_REMOVE(&head, data_ptr, socket_node, node_count);
+                free(data_ptr);
+                data_ptr = NULL;
             }
-        } while (read_bytes > 0);
-
-        close(file_fd);
-        if (SUCCESS == close(connection_fd))
-        {
-            syslog(LOG_INFO, "Closed connection from %s", ClientIpAddr);
         }
     }
+
+exit:
     close_app();
+    /* destroy mutex */
+    pthread_mutex_destroy(&thread_mutex);
```

Comment:
@chchbuff what does it mean to destroy the `thread_mutex` before joining threads?  How can you be certain this mutex won't be accessed after destroying?

Reply:
I am performing pthread_join for all sockets threads before destroy call. Below pthread_join is only for timer_thread.